src/main/java/com/slack/kaldb/metadata/core/CacheableMetadataStore.java

```diff
@@ -0,0 +1,101 @@
+package com.slack.kaldb.metadata.core;
+
+import com.slack.kaldb.metadata.zookeeper.CachedMetadataStore;
+import com.slack.kaldb.metadata.zookeeper.CachedMetadataStoreListener;
+import com.slack.kaldb.metadata.zookeeper.MetadataStore;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.state.ConnectionState;
+import org.slf4j.Logger;
+
+/**
+ * This store adds a cache on top of the KaldbMetadataStore. Optionally, this class can disable
+ * caching by passing in a feature flag.
+ *
+ * <p>Instantiating an instance of this class creates the cache, adds a listener and starts the
+ * cache. The listener doesn't notify the watchers on cache initialization but fires on subsequent
+ * cache events. To close the cache, call the close method.
+ */
+public abstract class CacheableMetadataStore<T extends KaldbMetadata>
```

Comment:
this name is super confusing as we already have CachedMetadataStore.

I think we should rename CachedMetadataStore to something else ? CachedMetadataStoreImpl  can just be CachedZKNode or something?

CacheableMetadataStore and CachedMetadataStore is tough to distinguish otherwise IMO

Reply:
Yes, I will prefix all the classes in zookeeper package with ZK. I haven't done that yet to keep the changes to a minimum. I can change it in a separate diff once this is merged.

Reply:
sounds good!